lib/actions-oclif/env/rm.ts

```diff
@@ -50,6 +50,7 @@ export default class EnvRmCmd extends Command {
 `;
 	public static examples = [
 		'$ balena env rm 123123',
+		'$ balena env rm 123123,234234',
```

Comment:
Oh wait, what I suggested isn't completely right either, as the env vars could still belong to multiple applications, devices or services, as long as they are of the same type (the types being app vars, device vars, service vars or config vars). Perhaps:
```
public static description = stripIndent`
    Remove config or env vars from applications, devices or services.
```
I think a note would have to be added to the help output to clarify that the multiple listed variables need to be of the same type. I assume that a variable ID like 123123 could refer to multiple different variables of different types, so that the disambiguation requires the pair (variable ID, variable type). If the implementation wanted to be cleverer, perhaps it could query all var types for a given variable ID. If only one variable existed for that ID (across all var types), it could be assumed to be the right one; but if multiple variables existed for the same ID (with different var types), then the user would need to disambiguate. It sounds a bit convoluted, and it users could be caught by surprise if they tested and scripted a command like `balena env rm 123123` that would "usually" work, but one day could fail with some "ambiguous variable ID error" (because some other user / application created a variable of a different type that by chance happened to have the same variable ID). I lean towards being deterministic, and always failing with an error message if a given variable ID cannot be found for the given (mandatory) variable type flag (`--device`, `--config`, `--service`, or flag absence meaning "application"). Thus insisting that all listed variable IDs (in the comma-separated list) must be for variables of the same type.

Reply:
Good catch! I did flip flop these from what was described in the issue. I will say that I am partial to the behavior being what was originally described in the issue in which there is one variable specified for multiple applications, devices, or services. This would make the behavior match `balena env add`. I think it would also fix the problem of different variable types if only one of application, device, or service was allowed at a time.

The problem I see with this however, is that I think it is rare for variables to share an id. I created a number of variables when testing `balena env add` and none of them share the same id, even the same variables created with bulk command. Having to look up each individual variable id almost negates the benefit of a bulk command IMO. 

A better solution I think would be if we allowed the user to specify a variable name and applications or devices with a services option(the same as `balena env add`). We could then do a lookup for the different device ids by doing something similar to `balena env`. This is speculation on my part that the necessary framework is all available but I think it would make for the most consistent and useful user experience.

Reply:
> _Having to look up each individual variable id almost negates the benefit of a bulk command IMO._ [...]
> _A better solution I think would be if we allowed the user to specify a variable name and applications or devices with a services option (the same as `balena env add`)._ [...] _I think it would make for the most consistent and useful user experience._

I agree. 👍   On this topic, using "raw" database IDs to identify any resources is a discouraged pattern / bad design as it makes it harder to do database migrations. UUIDs are OK, names are OK, but raw database IDs are not. It is an objective to change the CLI to avoid the use of raw database IDs, and this aligns with the preferred solution you describe.

> This is speculation on my part that the necessary framework is all available

It may actually be available. Looking superficially at the Node SDK reference:
https://www.balena.io/docs/reference/sdk/node-sdk/

If you search the reference page for "var" and look at the several ".remove" methods, it seems they allow removing variables by name, given an app / device / service ID. I didn't check that all cases are covered; I leave this to you if you are interested! :-)

Also, we'd have to discuss the command line format / options. If a breaking change is necessary, we could put it behind a feature switch like `isV12()` in [version.ts](https://github.com/balena-io/balena-cli/blob/v12.9.8/lib/utils/version.ts#L27) (adding a similar `isV13()` function). For example, without having given this much thought, a `--byname` option could be added to `balena env rm`. If either `--byname` is given or `isV13()` is true, then the new behavior applies. A solution that avoided a breaking change and avoided the need for flags might be even better, if possible and still clean. Parsing the variable name/ID to detect whether it is a numeric ID or non-numeric name may not be good enough, if a number could be used as a variable name (I guess a variable like '123' is allowed).

Reply:
`'$ balena env rm --byname EDITOR --application MyApp',`
`'$ balena env rm --byname EDITOR --application MyApp,MyApp2',`
`'$ balena env rm --byname EDITOR --application MyApp --service MyService',`
`'$ balena env rm --byname EDITOR --application MyApp,MyApp2 --service MyService,MyService2',`
`'$ balena env rm --byname EDITOR --application MyApp,MyApp2 --service MyService,MyService2 --yes',`
`'$ balena env rm --byname EDITOR --device 7cf02a6',`
`'$ balena env rm --byname EDITOR --device 7cf02a6,d6f1433',`
`'$ balena env rm --byname EDITOR --device 7cf02a6 --service MyService',`
`'$ balena env rm --byname EDITOR --device 7cf02a6,d6f1433 --service MyService,MyService2',`
`'$ balena env rm --byname EDITOR --device 7cf02a6,d6f1433 --service MyService,MyService2 --yes',`

My thoughts are to setup the command line arguments very similar to `env add`. A comment could be added to the description mentioning that `--byname` can be omitted when using version 13.0.0 or higher. 

Looking through the sdk, I only see a method of removing a serviceVar from an application missing. I do see options for configVars but I'm not sure how these fit into everything. Such as are they similar to service in that adding the service flag adds the var to the sub service of the application or device. The command line examples above would need to have the config option added to them.

I also need to look up some examples of how the `isV13()` will work. I'm planning to work on this over the weekend as I ran out of time this evening. I plan on following a similar code structure to `env add`.

Reply:
Oh... Let's think about this more carefully, especially the flags. It could get too messy, especially the fact that the `--device` and the `--service` flags would be either `boolean` or `string` flags depending on whether or not `--byname` was provided. Some ugly glue might be needed in the code, but most importantly, the documentation (`help`) would get complicated for users, as the flags (device, service) would mean different things depending on whether `--byname` was used. And the list of examples would have to include both raw numeric IDs and variable names, which makes it harder to understand either case as well.

Now I am thinking whether a new command would be preferable, like `balena env rm-by-name. Then, in CLI v13, `env rm-by-name` would be renamed `env rm` (with `env rm-by-name` being an alias for `env rm`). In this case, this PR would also have to add a deprecation / warning message to the current `env rm`, to warn users of the plans for v13.

Let's get a 3rd opinion. @srlowe, what do you think?

And, as the solution gets more involved, dare I ask, why are we doing all this, is it really needed? :-)  I think there is reasonable justification, not only for the added functionality of removing a variable from multiple devices / apps / services in a single CLI command, but also in relation to the desire of avoiding exposing raw balenaCloud database IDs to CLI users. I guess @srlowe would agree: He wrote "totally agree" on issue #1843. Actually, this PR would be (partially) addressing two issues, #1957 and #1843.

> Looking through the sdk, I only see a method of removing a serviceVar from an application missing.

If the SDK is missing functionality, then the right thing to do may be to add it as a PR to https://github.com/balena-io/balena-sdk first. Presumably the balena HTTP API provide the required method; one of the following? - https://www.balena.io/docs/reference/api/resources/application_environment_variable/  There is also precedent for the CLI querying the HTTP API directly. In fact, the present implementation of `balena env rm` does just that, using the Pine class:
```
 		await balena.pine.delete({ ... });
```

> I do see options for configVars but I'm not sure how these fit into everything. Such as are they similar to service in that adding the service flag adds the var to the sub service of the application or device.

If this helps:

```
$ balena help env add
...
  'BALENA_' or 'RESIN_' are reserved variable name prefixes used to identify
  "configuration variables". Configuration variables control balena platform
  features and are treated specially by balenaOS and the balena supervisor
  running on devices. They are also stored differently in the balenaCloud API
  database. Configuration variables cannot be set for specific services,
  therefore the --service option cannot be used when the variable name starts
  with a reserved prefix. When defining custom application variables, please
  avoid the reserved prefixes.
```
And the HTTP API resources:
* [Fleet environment variable](https://www.balena.io/docs/reference/api/resources/application_environment_variable/)
* [Fleet service variable](https://www.balena.io/docs/reference/api/resources/service_environment_variable/)
* [Fleet config variable](https://www.balena.io/docs/reference/api/resources/application_config_variable/)

In this context, I believe that the word "fleet" is used somewhat interchangeably with "application". In fact:

```
$ balena help env rm
...
  Variables are selected by their database ID (as reported by the 'balena envs'
  command) and one of six database "resource types":

  - application (fleet) environment variable
  - application (fleet) configuration variable (--config)
  - application (fleet) service variable (--service)
  - device environment variable (--device)
  - device configuration variable (--device --config)
  - device service variable (--device --service)

  The --device option selects a device-specific variable instead of an application
  (fleet) variable.

  The --config option selects a configuration variable. Configuration variable
  names typically start with the 'BALENA_' or 'RESIN_' prefixes and are used to
  configure balena platform features.

  The --service option selects a service variable, which is an environment variable
  that applies to a specifc service (application container) in a microservices
  (multicontainer) application.

  The --service and --config options cannot be used together, but they can be
  used alongside the --device option to select a device-specific service or
  configuration variable.
```

Reply:
I have some code for this that I am working on testing. I wanted to ping @srlowe for his thoughts how the command line args should be handled.

Reply:
Sorry guys, missed the ping.  My thoughts:

Looks like this is a lot more complex that it seems at first glance.

I'm very much for getting rid of the DB IDs in favour of names.  (But I'm not very keen on `--byname`.  Wouldn't it be better to deprecate support for IDs in a major version, and then temporarily support '--byid'?)  However, when specifying envs by name rather than by id, then the flags `-d` `-s` would have to change from boolean to string, to ensure that we don't remove stuff by mistake.  And we'd need to add an `-a` flag, and remove the assumption that they meant an application.  (I haven't thought through the implications of `-c`.

I'm not very keen on introducing a new command unless it's really necessary.

> And, as the solution gets more involved, dare I ask, why are we doing all this, is it really needed? :-)

I have to say this crossed my mind too.  Is the ability to remove multiple env vars across multiple devices and multiple services and multiple apps (etc) in one line, really needed?  There's been internal discussion recently about value-for-complexity, and I feel like this is something that scores very low there.

IMO I think this might need some internal discussion with the product team, as this isn't a simple change, and I don't know everyone's views on all the implications here.

Reply:
> Is the ability to remove multiple env vars across multiple devices and multiple services and multiple apps (etc) in one line, really needed?

It is not a priority, for example looking at this [ordered backlog](https://github.com/balena-io/balena-cli/milestone/2). But this PR would go beyond this feature, also reducing the use of raw database IDs (#1843) -- which is also not a priority, but certainly a step in the right direction. And, this is an external contribution, so not strictly bound by what balena might think a priority, though PR reviewing takes some effort from the balena team too. :-)

> value-for-complexity, and I feel like this is something that scores very low there

The complexity comes from the fact that the command was originally written to use raw database IDs, and it is a breaking change to undo this. Breaking changes require "coordination" with feature switches and add some work when the time comes to release v13. Nonetheless, it is true, as a whole it doesn't score highly in that metric.

> I'm not very keen on introducing a new command unless it's really necessary.

Right - having an extra command to do the same thing (`rm` and `rm-by-name`) creates confusion to end users, and at the very least makes the CLI look "untidy". But the new command would be temporary, just as a bridge until v13. Which however then requires extra "coordination" when the time comes to release v13, and this hits the value-for-complexity argument.

By the way @nwneisen, what is your motivation behind the recent CLI PRs? :-)  Depending on the motivation, if you would not be gaining particularly much from removing multiple env vars in a single CLI command, you could consider looking for a different challenge in the [ordered backlog](https://github.com/balena-io/balena-cli/milestone/2).

Reply:
I'll hold onto the changes I made in a branch on my forked repo but I will close this PR. If it comes up as more of a desired change then we can revisit how everything should be done then.

Reply:
@pdcastro Nothing specific for myself in these PRs just looking to get involved in the project. The original issue was to be able to move multiple devices in one command and these env commands were mentioned as additional changes that would be nice to have. See #323.

Reply:
@nwneisen By the way, just wanted to say thanks for taking the time/energy on this, it's appreciated!

Reply:
@srlowe I'm glad to be able to help. I like the Balena OS project and it's awesome to be able to work on it. I'll be looking for some higher priority issues to work on.